File: zorro_core/context/resolver.py
from enum import Enum
from typing import Dict, Optional, Set, List
from collections import defaultdict
from copy import deepcopy

from .plugin import Plugin
from zorro_core.main.config import PluginConfig
from zorro_core.utils.logger import logger


class VersionOperator(Enum):
    EQUAL = "=="
    LESS_EQUAL = ">="
    MORE_EQUAL = "<="


class VersionQuery:
    """
    Interpretation of a version query used in cases like plugin requirement.
    The query strings can look like myplugin>=2.0.3
    """

    OPERATORS_MAPPING: Dict[VersionOperator, str] = {
        VersionOperator.EQUAL: "==",
        VersionOperator.MORE_EQUAL: ">=",
        VersionOperator.LESS_EQUAL: "<=",
    }

    def __init__(self, query: str):
        self.name = query
        self.version = ""
        self.operator = self.OPERATORS_MAPPING[VersionOperator.EQUAL]

        for operator, token in self.OPERATORS_MAPPING.items():
            if len(query_split := query.split(token)) == 2:
                self.operator = operator
                self.name, self.version = query_split

    def match(self, plugin: Plugin) -> bool:
        """Test if the given plugin satisfies the query"""

        if self.operator == VersionOperator.EQUAL:
            return self.version == plugin.version
        if self.operator == VersionOperator.LESS_EQUAL:
            return plugin <= self
        elif self.operator == VersionOperator.MORE_EQUAL:
            return plugin >= self

        logger.warning("Could not match with version query %s: Invalid operator", self)
        return False


async def get_all_plugin_versions(name: str, config: PluginConfig) -> Set[Plugin]:
    """Find all available plugin versions in the current context"""

    plugins = set()
    for plugin_search_path in config.plugin_paths:
        # TODO: Customise the glob to stop the recursion when finding a zorro-plugin.json
        for plugin_path in plugin_search_path.glob("**/zorro-plugin.json"):
            plugin = await Plugin.load_bare(plugin_path)
            if plugin.name == name:
                plugins.add(plugin)

    return plugins


async def get_matching_plugins_from_query(
    query: str, config: PluginConfig
) -> Dict[str, Set[Plugin]]:
    """
    Get all available plugins that could be potential candidates to satisfy
    the query
    """

    plugin_versions = defaultdict(set)
    for single_query in query.split(" "):
        version_query = VersionQuery(single_query)

        for plugin in await get_all_plugin_versions(version_query.name, config):
            if version_query.match(plugin):
                plugin_versions[version_query.name].add(plugin)

    return plugin_versions


def get_prefered_plugin_version(plugins: Set[Plugin]) -> Optional[Plugin]:
    """
    When multiple plugin versions are potential quantidates, we use the
    lasted version of them.
    """

    if len(plugins) == 0:
        return None

    prefered_plugin = list(plugins)[0]
    for plugin in plugins:
        if plugin > prefered_plugin:
            prefered_plugin = plugin

    return prefered_plugin


def _combine_quandidates(
    quandidates_a: Dict[str, Set[Plugin]], quandidates_b: Dict[str, Set[Plugin]]
):
    """
    Create a new quandidate set with only quandidates that are present in both sets.
    """

    combined_quandidates: Dict[str, Set[Plugin]] = {}
    for key in set([*quandidates_a.keys(), *quandidates_b.keys()]):
        if key not in quandidates_a.keys():
            combined_quandidates[key] = quandidates_b.get(key, set())
        elif key not in quandidates_b.keys():
            combined_quandidates[key] = quandidates_a.get(key, set())
        else:
            combined_quandidates[key] = quandidates_a.get(key, set()).intersection(
                quandidates_b.get(key, set())
            )

    return combined_quandidates


async def _resolve_plugin(
    plugin_name: str,
    quandidates: Dict[str, Set[Plugin]],
    config: PluginConfig,
):
    print("RESOLVE")
    print(plugin_name)
    print([i.version for i in quandidates[plugin_name]])
    plugin = get_prefered_plugin_version(quandidates[plugin_name])
    if plugin is None:
        logger.error(
            "Could not resolve plugin %s: No valid versions available", plugin_name
        )
        return None
    print(plugin)

    # Make sure the plugin is fully loaded
    loaded_plugin = await plugin.reload()

    # Add all the requirement possible quandidates
    new_quandidates = quandidates
    for requirement in loaded_plugin.require:
        requirement_quandidates = await get_matching_plugins_from_query(
            requirement, config
        )
        print("REQUIREMENT")
        print(
            [
                {key: [i.version for i in value]}
                for key, value in requirement_quandidates.items()
            ]
        )
        print("ORIGINAL")
        print(
            [
                {key: [i.version for i in value]}
                for key, value in new_quandidates.items()
            ]
        )
        # We don't want to keep the quandidates does not match
        # the current requirements
        new_quandidates = _combine_quandidates(new_quandidates, requirement_quandidates)
        print("COMBINED")
        print(
            [
                {key: [i.version for i in value]}
                for key, value in new_quandidates.items()
            ]
        )

    # The prefered plugin's requirement might not be compatible with
    # the currently selected quandidates
    if any(len(quandidates) == 0 for quandidates in new_quandidates.values()):
        return None

    quandidates.update(new_quandidates)
    quandidates[plugin_name] = set([plugin])
    return plugin


async def _resolve_next_plugin_graph_iteration(
    quandidates: Dict[str, Set[Plugin]],
    config: PluginConfig,
    completed: Optional[Set[str]] = None,
):
    """
    Recusive function that will select a quandidates and resolve its dependencies.
    It will try every possible versions of the plugin until a valid combinason is found
    """

    completed = completed or set()
    # Used in case we need to try again with a different version choice
    original_quandidates = deepcopy(quandidates)

    # Select the next plugin that needs to be resolved
    plugin_name = next(
        (
            plugin_name
            for plugin_name in quandidates.keys()
            if plugin_name not in completed
        ),
        None,
    )

    # There is not plugins to resolve anymore, the resolution is complete
    if plugin_name is None:
        return True

    resolved_plugin = await _resolve_plugin(plugin_name, quandidates, config)
    # A plugin might be required but no quandidates
    # are valid anymore, this means that the graph we
    # are trying to resolve is impossible
    if resolved_plugin is None:
        return False

    completed.add(plugin_name)

    # Continue to resolve the next plugins,
    # the result will tell us if the resolution path was possible
    while not await _resolve_next_plugin_graph_iteration(
        quandidates, config, completed
    ):
        # Remove the choice we just made from the quandidates
        # until we make a choice that result in a possible resolution
        original_quandidates[plugin_name].remove(resolved_plugin)
        quandidates.update(deepcopy(original_quandidates))
        resolved_plugin = await _resolve_plugin(plugin_name, quandidates, config)
        if resolved_plugin is None:
            return False
    else:
        # The path chosen completed the requirements
        return True


async def resolve_plugins(query: str, config: PluginConfig) -> List[Plugin]:
    """
    Resolve a flat list of plugin that satisfies the given query
    """

    plugin_quandidates = await get_matching_plugins_from_query(query, config)
    if not await _resolve_next_plugin_graph_iteration(plugin_quandidates, config):
        logger.error("Could not resolve the depencency graph for the query %s", query)
        return set()

    return [versions.pop() for versions in plugin_quandidates.values()]

File: internal/context/resolver.go
package context

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/Acedyn/zorro-core/internal/config"
	"github.com/Acedyn/zorro-core/internal/utils"

	"github.com/life4/genesis/slices"
)

// List of all possible version operators
type VersionOperator string

const (
	VersionOperator_EQUAL      VersionOperator = "=="
	VersionOperator_LESS_EQUAL VersionOperator = "<="
	VersionOperator_MORE_EQUAL VersionOperator = ">="

	VERSION_ITEM_SEPARATOR = "."
	PLUGIN_DEFINITION_NAME = "zorro-plugin"
)

// Compaire two versions and define if the second one is less or more than the first
func CompareVersions(versionA string, versionB string) VersionOperator {
	// Handle the edge cases first
	if versionA == versionB {
		return VersionOperator_EQUAL
	} else if versionA == "" {
		return VersionOperator_LESS_EQUAL
	} else if versionB == "" {
		return VersionOperator_MORE_EQUAL
	}

	splittedVersionA := strings.Split(versionA, VERSION_ITEM_SEPARATOR)
	splittedVersionB := strings.Split(versionB, VERSION_ITEM_SEPARATOR)
	minVersionLenght, _ := slices.Min([]int{len(splittedVersionA), len(splittedVersionB)})

	// We compare the version items by items
	for index := 0; index < minVersionLenght; index++ {
		versionItemA := splittedVersionA[index]
		versionItemB := splittedVersionB[index]

		// If the item is a number the comparison is different
		versionNumberA, errA := strconv.Atoi(versionItemA)
		versionNumberB, errB := strconv.Atoi(versionItemB)

		// For number comparison, compare the values
		if errA == nil && errB == nil {
			if versionNumberA > versionNumberB {
				return VersionOperator_MORE_EQUAL
			} else if versionNumberA < versionNumberB {
				return VersionOperator_LESS_EQUAL
			}
			// For string comparison, compare aphabetically
		} else {
			switch strings.Compare(versionItemA, versionItemB) {
			case 1:
				return VersionOperator_MORE_EQUAL
			case -1:
				return VersionOperator_LESS_EQUAL
			}
		}
	}

	return VersionOperator_EQUAL
}

// Parsed version of a version query
type VersionQuery struct {
	Name     string
	Version  string
	Operator VersionOperator
}

// Test if the given plugin satisfies the query
func (versionQuery *VersionQuery) Match(plugin *Plugin) bool {
	versionComparison := CompareVersions(plugin.GetVersion(), versionQuery.Version)

	if versionComparison == VersionOperator_EQUAL {
		return true
	}
	if versionComparison == versionQuery.Operator {
		return true
	} else {
		return false
	}
}

func ParseVersionQuery(query string) *VersionQuery {
	versionQuery := VersionQuery{
		Name:     query,
		Version:  "",
		Operator: VersionOperator_EQUAL,
	}

	operators := []VersionOperator{VersionOperator_EQUAL, VersionOperator_LESS_EQUAL, VersionOperator_MORE_EQUAL}
	for _, operator := range operators {
		querySplit := strings.Split(query, string(operator))
		if len(querySplit) == 2 {
			versionQuery.Name = querySplit[0]
			versionQuery.Version = querySplit[1]
			versionQuery.Operator = operator
		}
	}

	return &versionQuery
}

// Find all available plugin versions with the given name
func FindPluginVersions(name string, pluginConfig *config.PluginConfig) []*Plugin {
	if pluginConfig == nil {
		pluginConfig = config.AppConfig().PluginConfig
	}
	versions := []*Plugin{}

	for _, pluginSearchPath := range pluginConfig.PluginPaths {
		err := filepath.WalkDir(pluginSearchPath, func(path string, f os.DirEntry, _ error) error {
			pathStem := strings.TrimSuffix(f.Name(), filepath.Ext(f.Name()))
			if pathStem == PLUGIN_DEFINITION_NAME {
				plugin := LoadPluginBare(path)
				if plugin.GetName() == name {
					versions = append(versions, LoadPluginBare(path))
				}
				return filepath.SkipDir
			}
			return nil
		})
		if err != nil {
			utils.Logger().Warn("An error occured while looking for plugins in path %s:\n\t%s", pluginSearchPath, err)
		}
	}

	return versions
}

// Get all available plugins that could be potential candidates to satisfy the query
func GetQueryMatchingPlugins(queries []string, pluginConfig *config.PluginConfig) map[string][]*Plugin {
	pluginVersion := map[string][]*Plugin{}
	groupedQueries := map[string][]*VersionQuery{}
	for _, query := range queries {
		versionQuery := ParseVersionQuery(query)
		if queryGroup, ok := groupedQueries[versionQuery.Name]; ok {
			groupedQueries[versionQuery.Name] = append(queryGroup, versionQuery)
		} else {
			groupedQueries[versionQuery.Name] = []*VersionQuery{versionQuery}
		}
	}

	for pluginName, versionQueries := range groupedQueries {
		if _, ok := pluginVersion[pluginName]; !ok {
			pluginVersion[pluginName] = []*Plugin{}
		}

		for _, plugin := range FindPluginVersions(pluginName, pluginConfig) {
			isMatch := true
			for _, versionQuery := range versionQueries {
				if !versionQuery.Match(plugin) {
					isMatch = false
				}
			}
			if isMatch {
				pluginVersion[pluginName] = append(pluginVersion[pluginName], plugin)
			}
		}
	}

	return pluginVersion
}

// When multiple plugin versions are potential quantidates, we use the
// lasted version of them.
func GetPreferedPluginVersion(versions []*Plugin) *Plugin {
	if len(versions) == 0 {
		return nil
	}

	preferedIndex := 0
	for pluginIndex, plugin := range versions {
		preferedPlugin := versions[preferedIndex]
		versionComparison := CompareVersions(plugin.GetVersion(), preferedPlugin.GetVersion())

		switch versionComparison {
		// If the version is higher the plugin take the spot of the prefered plugin
		case VersionOperator_MORE_EQUAL:
			preferedIndex = pluginIndex
		// For equal versions the more precise one win
		case VersionOperator_EQUAL:
			splittedPreferedVersion := strings.Split(preferedPlugin.GetVersion(), VERSION_ITEM_SEPARATOR)
			splittedVersion := strings.Split(plugin.GetVersion(), VERSION_ITEM_SEPARATOR)
			if len(splittedVersion) > len(splittedPreferedVersion) {
				preferedIndex = pluginIndex
			}
		}
	}

	return versions[preferedIndex]
}

// Create a new quandidate set with only quandidates that are present in both sets.
func intersectQuandidates(quandidatesA, quandidatesB map[string][]*Plugin) map[string][]*Plugin {
	intersectedCandidates := map[string][]*Plugin{}

	// Get all the keys into a set
	pluginsKeys := make(map[string]bool, len(quandidatesA))
	for key := range quandidatesA {
		pluginsKeys[key] = true
	}
	for key := range quandidatesB {
		pluginsKeys[key] = true
	}

	for key := range pluginsKeys {
		quandidateSetA, okA := quandidatesA[key]
		quandidateSetB, okB := quandidatesB[key]
		// The two first cases are simple: only one of the two has
		// plugins in the key so no intersections to do
		if !okA {
			intersectedCandidates[key] = quandidateSetB
		} else if !okB {
			intersectedCandidates[key] = quandidateSetA
		} else {
			// Both quandidates sets have plugins, we must only keep the ones
			// That are present in both
			intersectedCandidates[key] = []*Plugin{}

			for _, quandidateA := range quandidateSetA {
				for _, quandidateB := range quandidateSetB {

					nameEqual := quandidateA.GetName() == quandidateB.GetName()
					versionEqual := quandidateA.GetVersion() == quandidateB.GetVersion()
					pathEqual := quandidateA.GetPath() == quandidateB.GetPath()
					labelEqual := quandidateA.GetLabel() == quandidateB.GetLabel()

					if nameEqual && versionEqual && pathEqual && labelEqual {
						// Since they are equal, we don't care if we put the quandidate A or B
						intersectedCandidates[key] = append(intersectedCandidates[key], quandidateA)
						continue
					}
				}
			}
		}
	}

	return intersectedCandidates
}

// Select a plugin version from quandidates and add its dependencies to the quandidates
func resolvePluginVersion(
	name string,
	quandidates map[string][]*Plugin,
	pluginConfig *config.PluginConfig,
) (*Plugin, map[string][]*Plugin, error) {
	versions, ok := quandidates[name]
	if len(versions) <= 0 || !ok {
		return nil, nil, fmt.Errorf("no quandidates available for plugin %s", name)
	}

	// Reload the plugin to make sure it's not bare
	preferedVersion := GetPreferedPluginVersion(versions)
	plugin, error := LoadPluginFromFile(preferedVersion.GetPath())
	if error != nil {
		return preferedVersion, nil, fmt.Errorf("could not load plugin %s: %w", name, error)
	}

	// We don't want to keep the quandidates that does not match
	// the current requirements
	requirementQuandidates := GetQueryMatchingPlugins(plugin.GetRequire(), pluginConfig)
	newQuandidates := intersectQuandidates(quandidates, requirementQuandidates)

	newQuandidates[name] = []*Plugin{plugin}
	return preferedVersion, newQuandidates, nil
}

// Recusive function that will select a quandidates and resolve its dependencies.
// It will try every possible combinason until a valid one is fund
func resolvePluginGraph(
	quandidates map[string][]*Plugin,
	completed map[string]bool,
	pluginConfig *config.PluginConfig,
) (map[string][]*Plugin, error) {
	if completed == nil {
		completed = map[string]bool{}
	}

	// Select the next plugin that needs to be resolved
	var pluginToResolve *string = nil
	for quandidateName := range quandidates {
		if _, ok := completed[quandidateName]; !ok {
			pluginToResolve = &quandidateName
		}
	}

	// There is not plugins to resolve anymore, the resolution is complete
	if pluginToResolve == nil {
		return quandidates, nil
	}

	completed[*pluginToResolve] = true
	testedVersions := map[string]bool{}

	// Try to resolve a different plugin version until a valid graph is resolved
	for len(quandidates[*pluginToResolve]) > 0 {
		selectedVersion, newQuandidates, iterErr := resolvePluginVersion(*pluginToResolve, quandidates, pluginConfig)

		// Mark the selected version as tested and remove it from the quandidates
		testedVersions[selectedVersion.GetVersion()] = true
		quandidates[*pluginToResolve] = slices.Filter(quandidates[*pluginToResolve], func(plugin *Plugin) bool {
			_, ok := testedVersions[plugin.GetVersion()]
			return !ok
		})

		if iterErr != nil {
			utils.Logger().Debug(fmt.Sprintf("Skipping plugin versions: could not load plugin\n\t" + iterErr.Error()))
			continue
		}

		// First check if the resolved plugin resulted in a valid graph
		isValidGraph := true
		for pluginName, quandidateVersions := range newQuandidates {
			if len(quandidateVersions) == 0 {
				utils.Logger().Debug(fmt.Sprintf("Skipping plugin version %s: invalid resolved graph (no valid quandidates for plugin %s)", selectedVersion.GetVersion(), pluginName))
				isValidGraph = false
				break
			}
		}
		if !isValidGraph {
			continue
		}

		// Continue to resolve the graph
		if newQuandidates, err := resolvePluginGraph(newQuandidates, completed, pluginConfig); err != nil {
			utils.Logger().Debug(fmt.Sprintf("Skipping plugin versions %s: no graph combinason could be resolved", selectedVersion.GetVersion()))
			continue
		} else {
			// The resolved graph is valid
			return newQuandidates, nil
		}
	}

	return nil, fmt.Errorf("invalid graph: no valid combinason could not found")
}

// Resolve a flat list of plugin that satisfies the given query
func ResolvePlugins(query []string, pluginConfig *config.PluginConfig) ([]*Plugin, error) {
	if pluginConfig == nil {
		pluginConfig = config.AppConfig().PluginConfig
	}

	initialQuandidates := GetQueryMatchingPlugins(query, pluginConfig)
	resolvedGraph, err := resolvePluginGraph(initialQuandidates, nil, pluginConfig)
	if err != nil {
		return nil, fmt.Errorf("Plugin graph resolution imposible for query %s: %w", query, err)
	}

	resolvedPlugins := []*Plugin{}
	for _, plugins := range resolvedGraph {
		resolvedPlugins = append(resolvedPlugins, plugins[0])
	}

	return resolvedPlugins, nil
}
